### ISSUE TITLE
Local suppression of errors raised by extra checks from -Xcheck-macros

### ISSUE THREAD

**Issue by cchantep** [lampepfl/dotty#13926]
As indicated in the [documentation](https://dotty.epfl.ch/docs/reference/metaprogramming/macros.html#scope-extrusion), some scope extrusion case like `'{ (x: Int) => ${ run('x); 1 } }` are valid, but if compiled with `-Xcheck-macros` it will anyway raise a fatal error.

Being able to locally suppress such error, not to have to globally disable `-Xcheck-macros` would be nicer.

**Comment by nicolasstucki** [comment]
The documentation states that scope extrusions are invalid. Specifically this case.

**Comment by cchantep** [comment]
Sorry but then "This is again phase correct, but will lead us into trouble" is quite unclear.

**Comment by nicolasstucki** [comment]
Then we will need to rephrase that sentence to make it clear.

The point is that any kind of scope extrusions can leed to unsound code. In particular a `run` should never capture expressions that where not created within it's context.

**Comment by cchantep** [comment]
Honestly, the current scope/ownership rules make mixing quotes `'{ .. }`/`Expr` and `Term`/`Tree` quickly unusable.

**Comment by cchantep** [comment]
An [example](https://github.com/cchantep/ReactiveMongo-BSON/blob/d7003213c2c97d8d5db4badbe6dd1cdf403eeec6/api/src/main/scala-3/MacroImpl.scala#L869) where it compiled and run fine, but where `-Xcheck-macros` would raise a fatal error:

```
scala.quoted.runtime.impl.ScopeException: Expression created in a splice was used outside of that splice.
Created in: api/src/main/scala-3/MacroImpl.scala:869 at column 23
Used in: api/src/main/scala-3/MacroImpl.scala:858 at column 55
Expr: doc


Creation stack:
	api/src/main/scala-3/MacroImpl.scala:869 at column 23
	api/src/main/scala-3/Macros.scala:43 at column 5


Use stack:
	api/src/main/scala-3/MacroImpl.scala:858 at column 55
	api/src/main/scala-3/Macros.scala:43 at column 5
```

**Comment by nicolasstucki** [comment]
It looks like you are missing a `Quotes` argument in `appendDocCall`.

```diff
- def appendDocCall(doc: Expr[BSONDocument]): Expr[MBuilder[BSONElement, Seq[BSONElement]]] = {
+ def appendDocCall(doc: Expr[BSONDocument])(using Quotes): Expr[MBuilder[BSONElement, Seq[BSONElement]]] = {
```

The check indicates that in
```scala
                      ${ appendDocCall('{ doc }) }
```
the `'{ doc }` was created within this `${ ... }`  but because you are capturing the outer `Quotes` in `appendDocCall`.

The `-Xcheck-macros` seems to have correctly detected this situation. The code will work but because of this inconsistency the compiler will need to perform extra work to fix it after the fact. This is an instance where the owners are not created correctly and need to be changed when splicing. This can lead to exponential behavior in the worst case.

**Comment by cchantep** [comment]
If you mean that syntax like `'{ ${ foo('{ v }) } }` should not be used at all, then : 

- the compiler need to refuse that (even without`-Xcheck-macros`),
- the alternatives need to be documented.

**Comment by nicolasstucki** [comment]
No, I mean that `foo` should have a `(using Quotes)` parameter clause because in some cases the lack of it may cause performance issues. The compiler is resilient to this issue but `-Xcheck-macros` will complain to let the user know.

This is the most common case for this failure and we could give a hint in the message if the `-Xcheck-macros`. Usually, a simple way to fix it is to move the local method outside of the enclosing method and see that there is a missing `Quotes`.

We also need to add better documentation in the macro tutorial.

**Comment by cchantep** [comment]
The `using Quotes` is in the enclosing function

**Comment by nicolasstucki** [comment]
Indeed, the example has a bug an `-Xcheck-macros` cough it.

**Comment by ckipp01** [comment]
@nicolasstucki I see you did update docs. Do you feel like those updates are enough to close this, or are more needed?
